Add more configuration for gateway endpoints model mapping

TomeHirata opened the pull request:
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/TomeHirata/mlflow/pull/19618?quickstart=1)

#### Install mlflow from this PR

```
# mlflow
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/19618/merge
# mlflow-skinny
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/19618/merge#subdirectory=libs/skinny
```

For Databricks, use the following command:

```
%sh curl -LsSf https://raw.githubusercontent.com/mlflow/mlflow/HEAD/dev/install-skinny.sh | sh -s pull/19618/merge
```

</p>
</details>

### Related Issues/PRs

n/a

### What changes are proposed in this pull request?

This PR introduces new properties to the mapping between model definition and endpoint such as weights and fallback order. It updates proto and store methods, but it shouldn't be breaking changes since this feature is not yet released.

### How is this PR tested?

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [x] Manual tests

<!-- Attach code, screenshot, video used for manual testing here. -->

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

### Release Notes

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

<!-- Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change. -->

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows
- [x] `area/gateway`: MLflow AI Gateway client APIs, server, and third-party integrations
- [ ] `area/prompts`: MLflow prompt engineering features, prompt templates, and prompt management
- [ ] `area/tracing`: MLflow Tracing features, tracing APIs, and LLM tracing functionality
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

<!-- Do not modify or remove any text inside the parentheses. Keep both checkboxes below. -->

- [x] Yes (this PR will be cherry-picked and included in the next patch release)
- [ ] No (this PR will be included in the next minor release)

TomeHirata committed:
Add GatewayEndpointModelConfig support for gateway endpoints

This commit introduces the GatewayEndpointModelConfig class, which encapsulates the configuration for models attached to gateway endpoints. It includes fields for model definition ID, linkage type, weight, and fallback order. The API and internal handling of gateway endpoints have been updated to utilize this new configuration structure, replacing the previous model definition ID approach. Additionally, relevant documentation and tests have been updated to reflect these changes.

Signed-off-by: Tomu Hirata <tomu.hirata@gmail.com>

TomeHirata committed:
lint

Signed-off-by: Tomu Hirata <tomu.hirata@gmail.com>

BenWilson2 reviewed (approved):
Are there validations that should be done in the backend for configurations that may have downstream issues with weight sharing (i.e., the sum of all weights should not exceed 1.0). Ignore this comment if you're planning on enforcing that in the front end.

TomeHirata committed:
Refactor Gateway Endpoint Model Attachment

This commit updates the handling of model attachments to gateway endpoints by replacing the previous model definition ID and weight parameters with a new GatewayEndpointModelConfig structure. This change enhances the clarity and organization of model configurations, which now include linkage type, weight, and fallback order. The API, internal logic, and relevant tests have been adjusted accordingly to support this new configuration format.

Signed-off-by: Tomu Hirata <tomu.hirata@gmail.com>

TomeHirata committed:
Merge branch 'master' into feat/mapping-config

TomeHirata committed:
Refactor model configuration handling in gateway endpoints

This commit updates the model configuration structure for gateway endpoints, replacing the previous model definition ID and weight parameters with a new GatewayEndpointModelConfig format. Additionally, the weight normalization logic in the TrafficRouteProvider has been improved to ensure weights sum to 1. Relevant tests have been updated to reflect these changes.

Signed-off-by: Tomu Hirata <tomu.hirata@gmail.com>

TomeHirata committed:
test

Signed-off-by: Tomu Hirata <tomu.hirata@gmail.com>
